Some Remaining Strategies from Axelrod's first tournament

marcharper opened the pull request:
I gave this the ol' college try and would like some feedback before proceeding. In particular, note the following:
* Not all of the strategies are available in the [Fortran source](http://www-personal.umich.edu/%7Eaxe/research/Software/CC/CC2.html) -- only those that made it into the second tournament unaltered appear to be available
* The strategies `UnnamedStrategy`, `SteinRapoport`, and `Graaskamp` are underspecified and still incomplete, but as complete as possible otherwise
* For `SteinRapoport` we need to do a Chi-squared test. Since the source is not available, I recommend we use e.g. `scipy`
* `UnnamedStrategy` may be not much like it's original -- there is little information about it anywhere that I can find
* The Fortran code for `Graaskamp` in the detection of TFT and Analogy (completely unknown to me) is particularly difficult to follow, but this seems a minor piece of the strategy
* More tests are needed for the various complex behaviors
* The incomplete strategies have not yet been added to the list of strategies (but are imported into `_strategies.py`)
* The docs are not yet updated (pending #299)

I'm fairly confident that `Downing`, `Nydegger`, and `TidemanChieruzzi` are basically correct (pending some additional tests perhaps).

For the second Axelrod tournament strategies, I recommend we look into wrapping the Fortran source. Translating from Fortran to Python is very time consuming and error-prone because of the stylistic differences (GOTOs running amuck!). I think now there is a reason no one else has successfully reproduced the first tournament. That said, if we get similar results with these strategies, perhaps that is good enough?

Anyone want to email Axelrod to see if the source for the first tournament exists still? Apparently the first tournament was in BASIC.

drvinceknight commented:
> For SteinRapoport we need to do a Chi-squared test. Since the source is not available, I recommend we use e.g. scipy

Fine by me.

> Anyone want to email Axelrod to see if the source for the first tournament exists still? Apparently the first tournament was in BASIC.

I'll email now.

marcharper commented:
>> For SteinRapoport we need to do a Chi-squared test. Since the source is not available, I recommend we use e.g. scipy

> Fine by me

Even still, we need to know what the program did when the C and D counts are zero -- Chi-squared doesn't behave well in that case. Also I presume the original strategy used a stripped down calculation. We're really just guessing.

drvinceknight commented:
> Even still, we need to know what the program did when the C and D counts are zero -- Chi-squared doesn't behave well in that case. Also I presume the original strategy used a stripped down calculation. We're really just guessing.

Not sure what we can do...

Also re the Fortran strategies, my initial reservation has gone. I'm all for the wrapper, if at some point someone wanted to translate: awesome, in the meantime the strategies are there. Forgive my noobism, they/the library would still run even if I don't have Fortran right?

marcharper commented:
Yes, these should run fine. I'll run some sample tournaments and see how closely the results match the first tournament.

marcharper commented:
So I ran a test tournament with the strategies as is and compared the Axelrod's published results.

![](http://s3.postimg.org/q9p8pco3n/Axelrod_First_boxplot.png)

From Graaskamp to Random it's fairly similar. Downing wins, and I suspect this is because I coded "Revised Downing", unable to find the original Downing. Stein clearly needs it's test for Random. I don't know why Davis and Grudger fare so well, nor why TFT and Tideman so poorly. (Tideman may have bugs that testing will reveal -- my understanding of the FORTRAN differs from Axelrod's description.)

OTOH, since the outcome clearly depends on the total set of strategies, we may not be able to completely reproduce the original tournament's results.

marcharper commented:
For reference, the original score results were:
* 2.52: TFT
* 2.50: Tideman
* 2.43: Nydegger
* 2.41: Grofman
* 2.405: Shubik
* 2.39: SteinRapoport
* 2.365: Grudger (FRIEDMAN)
* 2.36: Davis
* 2.05: Graaskamp
* 1.955: Downing
* 1.64: Feld
* 1.52: Joss
* 1.505: Tullock
* 1.41: Unnamed
* 1.38: Random

marcharper commented:
So I'm going to leave this as is for the moment. I think I need more eyes on `TidemanChieruzzi`, `Graaskamp`, and `SteinRapoport`.

If anyone wants to try their own version of the tournament, here's the strategy list:
```
    strategies = [
        axelrod.TitForTat(),
        axelrod.Random(),
        axelrod.Davis(),
        axelrod.Downing(),
        axelrod.Feld(),
        axelrod.Graaskamp(),
        axelrod.Grofman(),
        axelrod.Grudger(), # FRIEDMAN
        axelrod.Nydegger(),
        axelrod.Joss(),
        axelrod.Shubik(),
        axelrod.SteinRapoport(),
        axelrod.TidemanChieruzzi(),
        axelrod.Tullock(),
        axelrod.UnnamedStrategy()
    ]
```

drvinceknight commented:
> So I'm going to leave this as is for the moment. I think I need more eyes on TidemanChieruzzi, Graaskamp, and SteinRapoport.

As soon as I get a chance I'll take a look but not sure when that will be. Hopefully someone else has a chance :)

Shame that it's so hard to replicate... I suppose that's why the library is here...

marcharper commented:
In the interest of preserving the useful changes here, I moved Stein, Tideman, and Graaskamp out to [another branch](https://github.com/marcharper/Axelrod/tree/axelrod_first_2), and cleaned up what was left.

If it's deemed merge-able I'll update the docs  for the new and modified strategies.

marcharper commented:
Rebased and updated docs. This PR now contains just the new strategies Nydegger, RevisedDowning, UnnamedStrategy, and an updated Grofman.

drvinceknight commented on axelrod/strategies/axelrod_first.py:
```diff
@@ -47,6 +46,79 @@ def strategy(self, opponent):
 
 
+class RevisedDowning(Player):
+    """Revised Downing."""
```
Can you add a one liner here?

drvinceknight commented on axelrod/strategies/axelrod_first.py:
```diff
@@ -237,108 +388,31 @@ def strategy(self, opponent):
-class Eatherley(Player):
-    """
-    Strategy submitted to Axelrod's second tournament by Graham Eatherley.
+    Warning: This strategy is incomplete.
```
Could you add a couple more details here.

drvinceknight commented on docs/tutorials/further_topics/classification_of_strategies.rst:
```diff
@@ -30,7 +30,7 @@ Or indeed find out how many strategy have only use 1 turn worth of memory to
 
     >>> len([s for s in axl.strategies if s().classifier['memory_depth']==1])
-    13
+    12
```
Oh how did this happen? What's the strategy that is no longer memory one?

marcharper replied:
Grofman

drvinceknight replied:
Cool (not that they're not accurate but that you've found that out).

drvinceknight commented:
:+1:

meatballs commented:
Once the merge conflicts are resolved, this is a :+1: from me

marcharper commented:
Rebase worked without conflict.
